consensus/istanbul/backend/announce.go

```diff
@@ -155,9 +213,80 @@ func (sb *Backend) sendIstAnnounce() error {
 	enodeUrl := enode.String()
 	view := sb.core.CurrentView()
 
-	msg := &announceMessage{Address: sb.Address(),
-		EnodeURL: enodeUrl,
-		View:     view}
+	regVals, err := sb.retrieveRegisteredValidators()
+	// The validator contract may not be deployed yet.
+	// Even if it is deployed, it may not have any registered validators yet.
+	if err == errValidatorsContractNotRegistered || len(regVals) == 0 {
+		sb.logger.Trace("Can't retrieve the registered validators.  Only allowing the initial validator set to send announce messages", "err", err, "regVals", regVals)
+		valSet := sb.getValidators(block.Number().Uint64(), block.Hash())
+
+		regVals = make(map[common.Address]bool)
+		for _, val := range valSet.List() {
+			regVals[val.Address()] = true
+		}
+	} else if err != nil {
+		sb.logger.Error("Error in retrieving the registered validators", "err", err)
+		return err
+	}
+
+	useValidatorContract := true
+	var g core.Genesis
+	validatorsAddress, err := sb.regAdd.GetRegisteredAddressAtCurrentHeader(params.ValidatorsRegistryId)
+	if err != nil {
+		log.Warn("Registry address lookup failed", "err", err)
+		useValidatorContract = false
+		genesisJSON, err := sb.db.Get(core.DBGenesisKey)
+		err = g.UnmarshalJSON(genesisJSON)
+		if err != nil {
+			log.Error("Unable to retrieve genesis from db", "err", err)
+		}
+	}
+
+	encryptedEnodeUrls := make(map[common.Address][]byte)
+	for addr := range regVals {
+		var pubKeyBytes []byte
+		if useValidatorContract {
+			state, err := sb.stateAt(block.Header().ParentHash)
+			if err != nil {
+				log.Error("verify - Error in getting the block's parent's state", "parentHash", block.Header().ParentHash.Hex(), "err", err)
+				return err
+			}
+			var validator []interface{}
+			if _, err := sb.iEvmH.MakeStaticCall(*validatorsAddress, getValidatorFuncABI, "getValidator", []interface{}{}, &validator, uint64(1000000), block.Header(), state); err != nil {
+				log.Error("Unable to retrieve Validator Account from Validator smart contract", "err", err)
+				return err
+			}
+			pubKeyBytes, _ = validator[3].([]byte) // The publicKey field
+		} else {
+			pubKeyBytes = g.Alloc[addr].PublicKey
+		}
+		// ECDSAKey, err := crypto.UnmarshalPubkey(pubKeyBytes)
+		// if err != nil {
+		// 	log.Info("bigsad")
+		// 	continue;
+		// }
+		// pubKey := ecies.ImportECDSAPublic(ECDSAKey)
+		pubKey, err := p2p.ImportPublicKey(pubKeyBytes)
+		encryptedEnodeUrl, err := ecies.Encrypt(rand.Reader, pubKey, []byte(enodeUrl), nil, nil)
```

Comment:
I think it would be great if @kobigurk reviewed this piece.  I know that there could be some important ways randomness needs to be used with signing (which I assume encryption would have the same requirement), so having a crypto expert take a look at this would be great.

Reply:
The choice of crypto.rand as a randomness source is good - it uses OS randomness, it's thread-safe and is the go-to choice for many libraries.

I have some general comments about the design:
* I think the same key is used for signing and encryption (correct me if I'm wrong). This is not good key hygiene in general. I think this was something that was added when istanbul was implemented, and if I recall correctly, signing keys and node p2p keys are unrelated in vanilla Geth. Worth thinking more about.
* EIP8 introduced the ciphertext length as the S2 parameter. I want to recall why and if it's important for our use.
* Something @nicholasguo and I discussed once - if we're using the signing key as the p2p key (which I think is the current design), this makes using a hardware wallet harder.

Reply:
Maybe my information is out of date - do we eventually have distinct keys?
Currently the code, I believe, verifies that the enode public key derives a valid validator address.

Reply:
Thanks for the feedback @kobigurk !

To clarify, @nicholasguo 's current implementation has the following properties:

1) The istanbul consensus message signing key and enodeURL encryption key are the same.
2) The key in 1) would be able to be stored in a hardware wallet.
3) All p2p communication will be using the node key to encrypt (via the rlpx protocol).  This key def. should be different than the key in 1), but if a user configures their geth to have their node key be the same as key in 1), then we wouldn't stop them.

I hope that clarifies everything!

Reply:
Hey @kevjue,

Thanks! How can decryption be done without having access to the key in 1 outside of a hardware wallet? I think I'm missing something obvious.

Reply:
I looked again at the area of code I thought was deriving an address from the enodeURL, and it seems I was wrong, so I think there's no problem having the enodeURL different than the signing key, which is great.

Maybe a good compromise regarding 1 would be to have different keys for actual block signatures and "just" istanbul messages. If it makes sense, I'll also make the BLS code treat it like that.